### PR TITLE
Use libsecp256k1 for secio handshakes

### PR DESCRIPTION
**Pull request by dvdplm**
Experiment using `libsecp256k1` to get rid of plenty of `unsafe` and gain `no_std` capability.

Related to https://github.com/libp2p/rust-libp2p/issues/627

### CODE REVIEW

File: protocols/secio/src/handshake.rs [hunk 1]
```diff
@@ -490,15 +487,15 @@ where
...
+                    // TODO: is the key compressed or not?
+                    let remote_public_key = secp256k1::PublicKey::parse_slice(remote_public_key, None);
+
                     if let (Ok(signature), Ok(remote_public_key)) = (signature, remote_public_key) {
-                        match secp256k1.verify(&message, &signature, &remote_public_key) {
-                            Ok(()) => (),
-                            Err(_) => {
+                        match secp256k1::verify(&message, &signature, &remote_public_key) {
+                            true => (),
+                            false => {
```
Comment (twittner) [inline protocols/secio/src/handshake.rs]: Why not `if !secp256k1::verify(&message, &signature, &remote_public_key) { ... }`?

File: protocols/secio/src/lib.rs [hunk 2]
```diff
@@ -271,13 +271,11 @@ impl SecioKeyPair {
...
-        let secp = secp256k1::Secp256k1::new();
-        // TODO: This will work once 0.11.5 is released. See https://github.com/rust-bitcoin/rust-secp256k1/pull/80#pullrequestreview-172681778
-        // let private = secp256k1::key::SecretKey::new(&secp, &mut secp256k1::rand::thread_rng());
         use rand::Rng;
-        let mut random_slice= [0u8; secp256k1::constants::SECRET_KEY_SIZE];
+        const SECRET_KEY_SIZE: usize = 32; // TODO: coming to libsecp256k1 as well
+        let mut random_slice= [0u8; SECRET_KEY_SIZE];
         rand::thread_rng().fill(&mut random_slice[..]);
-        let private = secp256k1::key::SecretKey::from_slice(&secp, &random_slice).expect("slice has the right size");
+        let private = secp256k1::SecretKey::parse(&random_slice).expect("slice has the right size");
```
Comment (twittner) [inline protocols/secio/src/lib.rs]: Parsing can fail for reasons other than size. Why not `let private = secp256k1::SecretKey::parse(&random_slice)?;` instead?

**Comment by tomaka** [comment]
The fate of this PR basically depends on the maturity of the `libsecp256k1` library.

**Comment by dvdplm** [comment]
> The fate of this PR basically depends on the maturity of the `libsecp256k1` library.

Indeed. It has seen some recent updates and might be used in `parity-crypto` ([PR](https://github.com/paritytech/parity-common/pull/80)) so there's that.

**Comment by tomaka** [comment]
Considering that we don't want to use secp256k1 on the long term (cc @burdges), I think it's a non-bad idea to merge this PR.

**Comment by burdges** [comment]
We should ideally avoid using long-term secp256k1 keys in tight loops, like authenticating ourselves to numerous incoming connections, just because it's harder to make implementations of Weierstrass curves be constant-time.  

If anyone really needs this, like if Ethereum on libp2p wants this, then we can easily make this secure with key splitting, but at half speed.  We could also pay reviewers to evaluate libsecp256k1 vs blocksteem's C library vs the amcl library's claims.  Appears amcl just sketched out and pulled all their repos down.  lol

**Comment by tomaka** [comment]
Unfortunately the `core` now depends on `secp256k1` as well, and should be updated.

**Comment by dvdplm** [comment]
Closing in favour of #1029

**Pull request closed without merging**